### PR TITLE
Add optional Istio local gateway

### PR DESCRIPTION
**Pull request by kaxil**
- `istio-local-gateway.yaml` the file is produced by the following:

```bash
curl -L https://git.io/getLatestIstio | ISTIO_VERSION=1.3.0 sh -
mv istio-1.3.0 istio
cat istio/install/kubernetes/helm/istio/example-values/values-istio-gateways.yaml | sed -e "s/custom-gateway/cluster-local-gateway/g" -e "s/customgateway/clusterlocalgateway/g" > ./istio-local-gateway.yaml

helm template --namespace=istio-system \
  --set gateways.custom-gateway.autoscaleMin=1 \
  --set gateways.custom-gateway.autoscaleMax=1 \
  --set gateways.custom-gateway.cpu.targetAverageUtilization=60 \
  --set gateways.custom-gateway.labels.app='cluster-local-gateway' \
  --set gateways.custom-gateway.labels.istio='cluster-local-gateway' \
  --set gateways.custom-gateway.type='ClusterIP' \
  --set gateways.istio-ingressgateway.enabled=false \
  --set gateways.istio-egressgateway.enabled=false \
  --set gateways.istio-ilbgateway.enabled=false \
  istio/install/kubernetes/helm/istio -f istio-local-gateway.yaml > istio-local-gateway.yaml

```

### CODE REVIEW

**Comment by dimberman** [comment]
@sjmiller609 can you confirm we've done nothing stupid so we can merge? :)

**Comment by sjmiller609** [comment]
I'm talking about these "set"s
```
helm template --namespace=istio-system \
  --set gateways.custom-gateway.autoscaleMin=1 \
  --set gateways.custom-gateway.autoscaleMax=1 \
  --set gateways.custom-gateway.cpu.targetAverageUtilization=60 \
  --set gateways.custom-gateway.labels.app='cluster-local-gateway' \
  --set gateways.custom-gateway.labels.istio='cluster-local-gateway' \
  --set gateways.custom-gateway.type='ClusterIP' \
  --set gateways.istio-ingressgateway.enabled=false \
  --set gateways.istio-egressgateway.enabled=false \
  --set gateways.istio-ilbgateway.enabled=false \
  istio/install/kubernetes/helm/istio -f istio-local-gateway.yaml > istio-local-gateway.yaml
```
plus the content of istio-local-gateway.yaml being in a block of yaml in locals.tf, then using "values" attribute on the helm_release.istio resource

**Comment by dimberman** [comment]
I think we just ran into a lot of issues with the "replacement" part especially since this is code that should very rarely change. We weren't able to work around the need to pre-download the file for validation (at which point we might as well just have the file we're trying to generate).

File: istio.tf [hunk 1]
```diff
@@ -52,3 +52,19 @@ resource "helm_release" "istio" {
 
   values = [var.extra_istio_helm_values]
```
Comment (sjmiller609) [inline istio.tf]: right here you should add the block of helm values for the local gateway rather than applying spec manually after the fact.

Comment (kaxil) [reply istio.tf]: The 1st commit does exactly that:

https://github.com/astronomer/terraform-kubernetes-astronomer-system-components/pull/6/commits/eb2fe7f3477c0ff3f7ea8572831e786ddd64c4f4

so if we want we can revert to it if you and @dimberman thinks so.

**Comment by sjmiller609** [comment]
can't you just paste this block into locals.tf
```

gateways:
  enabled: true

  custom-gateway:
    enabled: true
    labels:
      app: custom-gateway
    replicaCount: 1
    autoscaleMin: 1
    autoscaleMax: 5
    resources: {}
      # limits:
      #  cpu: 100m
      #  memory: 128Mi
      #requests:
      #  cpu: 1800m
      #  memory: 256Mi
    cpu:
      targetAverageUtilization: 80
    loadBalancerIP: ""
    loadBalancerSourceRanges: {}
    externalIPs: []
    serviceAnnotations: {}
    podAnnotations: {}
    type: LoadBalancer #change to NodePort, ClusterIP or LoadBalancer if need be
    #externalTrafficPolicy: Local #change to Local to preserve source IP or Cluster for default behaviour or leave commented out
    ports:
      ## You can add custom gateway ports
    - port: 80
      targetPort: 80
      name: http2
      # nodePort: 31380
    - port: 443
      name: https
      # nodePort: 31390
    - port: 31400
      name: tcp
      # nodePort: 31400
    # Pilot and Citadel MTLS ports are enabled in gateway - but will only redirect
    # to pilot/citadel if global.meshExpansion settings are enabled.
    - port: 15011
      targetPort: 15011
      name: tcp-pilot-grpc-tls
    - port: 8060
      targetPort: 8060
      name: tcp-citadel-grpc-tls
    # Addon ports for kiali are enabled in gateway - but will only redirect if
    # the gateway configuration for the various components are enabled.
    - port: 15029
      targetPort: 15029
      name: http2-kiali
    # Telemetry-related ports are enabled in gateway - but will only redirect if
    # the gateway configuration for the various components are enabled.
    - port: 15030
      targetPort: 15030
      name: http2-prometheus
    - port: 15031
      targetPort: 15031
      name: http2-grafana
    - port: 15032
      targetPort: 15032
      name: http2-tracing
    secretVolumes:
    - name: customgateway-certs
      secretName: istio-customgateway-certs
      mountPath: /etc/istio/customgateway-certs
    - name: customgateway-ca-certs
      secretName: istio-customgateway-ca-certs
      mountPath: /etc/istio/customgateway-ca-certs

```

changing whatever you need to?

File: istio.tf (hunk 1 above)
Comment (sjmiller609) [reply istio.tf]: no that commit has kube spec in locals.tf I don't think any of these will work in practice, you should try deploying the module.

Comment (sjmiller609) [reply istio.tf]: it should have helm values yaml rather than kube spec yaml

**Comment by dimberman** [comment]
@sjmiller609 the problem is the `sed` commands. There doesn't seem to be any non-super-gross way to replace `custom-gateway` with `local-gateway` using the helm tools available.

**Comment by kaxil** [comment]
So we need to create resource from the file that is generated from an example file in Istio repo:

```
helm template --namespace=istio-system \
  --set gateways.custom-gateway.autoscaleMin=1 \
  --set gateways.custom-gateway.autoscaleMax=1 \
  --set gateways.custom-gateway.cpu.targetAverageUtilization=60 \
  --set gateways.custom-gateway.labels.app='cluster-local-gateway' \
  --set gateways.custom-gateway.labels.istio='cluster-local-gateway' \
  --set gateways.custom-gateway.type='ClusterIP' \
  --set gateways.istio-ingressgateway.enabled=false \
  --set gateways.istio-egressgateway.enabled=false \
  --set gateways.istio-ilbgateway.enabled=false \
  istio/install/kubernetes/helm/istio -f istio-local-gateway.yaml > istio-local-gateway.yaml
```

All of the above commands just generates a single file `istio-local-gateway.yaml` that we need to apply.

It is separate from istio I guess, right @dimberman ?

**Comment by sjmiller609** [comment]
> @sjmiller609 the problem is the `sed` commands. There doesn't seem to be any non-super-gross way to replace `custom-gateway` with `local-gateway` using the helm tools available.

it is not necessary to sed because you can edit the helm values in locals.tf by hand.

**Comment by kaxil** [comment]
@dimberman Can you please check and test locally on your machine too !! 

And remove things from `istio_local_gateway_helm_values` in `locals.tf` that are not needed please.

**Comment by dimberman** [comment]
@kaxil so if I understand this you're now setting up the local gateway at the same time as you're setting up istio?

**Comment by kaxil** [comment]
@dimberman Yes

File: istio.tf [hunk 2]
```diff
@@ -50,5 +50,36 @@ resource "helm_release" "istio" {
   version    = var.istio_helm_release_version
   wait       = true
 
-  values = [var.extra_istio_helm_values]
+  values = [
+    var.extra_istio_helm_values,
```
Comment (sjmiller609) [inline istio.tf]: Add a line above this line
```
var.istio_local_gateway_helm_values,
```

File: istio.tf [hunk 3]
```diff
@@ -50,5 +50,36 @@ resource "helm_release" "istio" {
...
+  ]
+
+  dynamic "set" {
+    for_each = var.istio_local_development ? map("gateways.istio-ingressgateway.type", "NodePort") : {}
+    content {
+      name  = set.key
+      value = set.value
+    }
+  }
```
Comment (sjmiller609) [inline istio.tf]: delete this block

File: istio.tf [hunk 4]
```diff
@@ -50,5 +50,36 @@ resource "helm_release" "istio" {
...
+
+  values = [local.istio_local_gateway_helm_values]
+
+  dynamic "set" {
+    for_each = var.enable_istio_local_gateway ? local.istio_local_gateway_helm_set_list : {}
+    content {
+      name  = set.key
+      value = set.value
+    }
+  }
```
Comment (sjmiller609) [inline istio.tf]: delete this block

File: locals.tf [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    "gateways.custom-gateway.autoscaleMin"                 = "1"
+    "gateways.custom-gateway.autoscaleMax"                 = "1"
+    "gateways.custom-gateway.cpu.targetAverageUtilization" = "60"
+    "gateways.custom-gateway.labels.app"                   = "cluster-local-gateway"
+    "gateways.custom-gateway.labels.istio"                 = "cluster-local-gateway"
+    "gateways.custom-gateway.type"                         = "ClusterIP"
+    "gateways.istio-ingressgateway.enabled"                = "false"
+    "gateways.istio-egressgateway.enabled"                 = "false"
+    "gateways.istio-ilbgateway.enabled"                    = "false"
+  }
```
Comment (sjmiller609) [inline locals.tf]: delete this block

File: variables.tf [hunk 6]
```diff
@@ -160,3 +160,13 @@ variable "tiller_node_selectors" {
...
+
+variable "enable_istio_local_gateway" {
+  type    = bool
+  default = false
+}
+
+variable "istio_local_development" {
+  default     = false
+  description = "When using kubernetes kind for creating cluster turn this to true"
+}
```
Comment (sjmiller609) [inline variables.tf]: Delete this block

File: locals.tf [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+tracing:
+  enabled: false
+
+kiali:
+  enabled: false
+
+certmanager:
+  enabled: false
+
```
Comment (sjmiller609) [inline locals.tf]: Delete this block, replace with
```
gateways:
  enabled: true
  cluster-local-gateway:
    enabled: true
    labels:
      app: cluster-local-gateway
      istio: cluster-local-gateway
    type: ClusterIP
  istio-ingressgateway:
    enabled: false
  istio-ilbgateway:
    enabled: false
  istio-egressgateway:
    enabled: false
```

Comment (kaxil) [reply locals.tf]: It removes the following components if only the block you have suggested is kept:

```
istio-system   istio-citadel-679b7c9b5b-g7fgx
istio-system   istio-galley-5b597b94bb-4rchr 
istio-system   istio-ingressgateway-664dd5f597-h4kjb
istio-system   istio-pilot-5d5966685f-v45ps
istio-system   istio-policy-65b7ff57cc-6pjh7
istio-system   istio-sidecar-injector-68f4668959-sblcc
istio-system   istio-telemetry-5988f98454-d6vs2
```

Comment (sjmiller609) [reply locals.tf]: Most of those should already be installed. For example, here is the current stage cloud:
```
15ab373b5bdf:  ~/shared/google-environments/stage/cloud 
_ kubectl get pods -n istio-system
NAME                                      READY   STATUS      RESTARTS   AGE
istio-citadel-b46584495-9lz2m             1/1     Running     0          7d4h
istio-galley-785975f449-dsh2n             1/1     Running     0          7d4h
istio-galley-785975f449-z64sw             1/1     Running     0          7d4h
istio-init-crd-10-1.3.0-6m58j             0/1     Completed   0          7d4h
istio-init-crd-11-1.3.0-lvvrc             0/1     Completed   0          7d4h
istio-init-crd-12-1.3.0-5vpxx             0/1     Completed   0          7d4h
istio-pilot-6f6ff4f579-2zvpb              2/2     Running     0          115m
istio-pilot-6f6ff4f579-rsdfc              2/2     Running     0          115m
istio-policy-5bb46bd75f-fvsdz             2/2     Running     0          115m
istio-policy-5bb46bd75f-xmc6m             2/2     Running     0          115m
istio-sidecar-injector-5c9588b458-9j2t5   1/1     Running     0          7d4h
istio-sidecar-injector-5c9588b458-sntc5   1/1     Running     0          7d4h
istio-telemetry-66d99df56-2njvb           2/2     Running     0          115m
istio-telemetry-66d99df56-pcblb           2/2     Running     0          115m
kiali-6ccf68bc58-849xg                    1/1     Running     0          7d4h
prometheus-65d7ff6dc4-7txgq               1/1     Running     0          7d4h
```

which configuration change causes them to be deleted?

Comment (sjmiller609) [reply locals.tf]: I recommend checking your helm values with "helm get" on your development set up, and make sure the final, merged configuration looks right.

File: locals.tf (hunk 5 above)
Comment (kaxil) [reply locals.tf]: Removed

**Comment by sjmiller609** [comment]
@kaxil let me know when you're ready for another review

**Comment by kaxil** [comment]
@sjmiller609 The PR is ready but for some reason we face a issue on @dimberman mac that I am not able to reproduce.

The issue is :

```
values = [var.enable_istio_local_gateway ? local.istio_local_gateway_helm_values : "", var.extra_istio_helm_values]
```
retuned ["....all..valeues", ""]

as `var.extra_istio_helm_values` is `""` it is giving some error on his machine

@dimberman - Can you post the stacktrace.

I have added `compact` (https://www.terraform.io/docs/configuration/functions/compact.html) that should have removed that empty string but he still gets error. It might have something to do with "kind" but don't know

```
values = compact([var.enable_istio_local_gateway ? local.istio_local_gateway_helm_values : "", var.extra_istio_helm_values])
```

File: locals.tf [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    resources: {}
+    secretVolumes:
+    - mountPath: /etc/istio/clusterlocalgateway-certs
+      name: clusterlocalgateway-certs
+      secretName: istio-clusterlocalgateway-certs
+    - mountPath: /etc/istio/clusterlocalgateway-ca-certs
+      name: clusterlocalgateway-ca-certs
+      secretName: istio-clusterlocalgateway-ca-certs
+    serviceAnnotations: {}
+    type: LoadBalancer
```
Comment (sjmiller609) [inline locals.tf]: is this block required?

Comment (kaxil) [reply locals.tf]: Yes !

Comment (kaxil) [reply locals.tf]: This is what I get from Helm get.

And I think all the values except defaults are required.

@dimberman can you check please if we don't need anything?

Comment (sjmiller609) [reply locals.tf]: Below this block there is another gateway. Are you intending to make two gateways or one?

File: locals.tf [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+  custom-gateway:
```
Comment (sjmiller609) [inline locals.tf]: this is a second gateway

File: locals.tf (hunk 8 above)
Comment (sjmiller609) [reply locals.tf]: 'helm get' just shows you the user defined and computed values.

File: locals.tf [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+      targetAverageUtilization: 60
+    labels:
+      app: cluster-local-gateway
+      istio: cluster-local-gateway
+    type: ClusterIP
+  enabled: true
+  istio-egressgateway:
+    enabled: false
+  istio-ilbgateway:
+    enabled: false
```
Comment (sjmiller609) [inline locals.tf]: this is a second gateway

File: locals.tf [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    type: LoadBalancer
+  custom-gateway:
+    autoscaleMax: 1
+    autoscaleMin: 1
+    cpu:
+      targetAverageUtilization: 60
+    labels:
+      app: cluster-local-gateway
+      istio: cluster-local-gateway
+    type: ClusterIP
```
Comment (sjmiller609) [inline locals.tf]: this is a second gateway

Comment (kaxil) [reply locals.tf]: Updated !

File: locals.tf (hunk 8 above)
Comment (kaxil) [reply locals.tf]: I agree I have updated to min required stuff.

Comment (kaxil) [reply locals.tf]: btw ports is a required field.

Without port it gives following error:

```
Error: rpc error: code = Unknown desc = release istio failed: Service "cluster-local-gateway" is invalid: spec.ports: Required value
```

Comment (kaxil) [reply locals.tf]: So I added the port too

File: locals.tf [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+      - name: https
+        port: 443
+    type: ClusterIP
+  enabled: true
+  istio-egressgateway:
+    enabled: false
+  istio-ilbgateway:
+    enabled: false
+  istio-ingressgateway:
+    enabled: false
```
Comment (kaxil) [inline locals.tf]: Not sure we are actually using `istio-ingressgateway` anywhere.

cc @schnie @sjmiller609

Comment (sjmiller609) [reply locals.tf]: that's right.

**Pull request merged**